Review PR398 計算結果に表示する項目を一元管理

kei-gbf opened the pull request:
- データ構造に配列を利用
  繰り返し時の順序を保証する為
- switch フラグの状態を 1 or 0 から true or false に変更
  各所で 数値→論理値変換のコードが必要になる為。
  ※この変更により __disableAutoResultUpdate のテスト__ も必要になります（影響範囲）
- テーブルを３列毎に分割する為、lodashライブラリのchunk互換関数を実装
  (但し実行効率は悪いが、小さな用途なので必要十分）
- Object.assign to combine default state 初期状態をswitchListから生成し Object.assign で合成
- Dropped "Chara" prefix
  モバイル版PC版双方でコードが重複。
  必要な場合は、switchList 側で一元管理した方が良い

極力、元のコードからの変更を少なくするようにしました。
スタイルシートは適応してないので、見栄え等は適宜変更してください。

kei-gbf committed:
Review PR398

- use Array as switchList data structure
  ensure the iteration order
- Iterate category of switchList
- Change flag state number to boolean
  because number had to convert to boolean everywhere
- porting lodash.chunk compatible function
- Object.assign to combine default state
- Dropped "Chara" prefix

kei-gbf commented on src/result.js:
```diff
@@ -41,48 +41,67 @@ var {
+        {name: "switchSkillTotal", label: "スキル合計"},
+        {name: "switchDebuffResistance", label: "弱体耐性率"},
+    ]},
+];
```
JSX表記は一旦断念し、配列で表現します。
- インデントを揃えたので diff (左右表示) で比較がしやすい（はず）
- 将来的に JSX への編集が簡単。(オブジェクトのプロパティが要素の属性に対応）

kei-gbf commented on src/result.js:
```diff
@@ -41,48 +41,67 @@ var {
+      ? [...arr, [item]]
+      : [...arr.slice(0, -1), [...arr.slice(-1)[0], item]];
+  }, []);
+};
```
リストの要素を分割する汎用関数。
リストをテーブルで n 列表示する際の定石です。

- 短いコードなので上URLよりコピーしました。ライセンスは同じMITなので参照元を記載すれば問題なし。
- 下URLには実行効率が悪い～的なことが書かれてますが、小規模なデータで適応する為それ程問題になりません。
- コード内容が一見して何してるかわからない点等はコメントで説明補足。

kei-gbf commented on src/result.js:
```diff
@@ -219,33 +238,29 @@ var ResultList = CreateClass({
+    getInitialSwitchState: function (_switchList=switchList) {
+        return Object.assign(..._switchList.map(({items}) =>
+            Object.fromEntries(items.map(({name, selected}) => [name, selected||false]))));
+    },
```
switchList から初期stateを生成する関数

kei-gbf replied:
コードを保守する上で、Array.map を用いた式は
読みにくいのが問題かもしれないので説明補足。
for 文で記述した互換コードは以下の通りです。

```js
const res = {};
for (const {items} of _switchList) {
    for (const {name, selected} of items) {
        res[name] = selected || false;
    }
}
```

kei-gbf replied:
読み返してみると
spread構文にObject.assign, Object.fromEntries, Array.map と乱用しすぎてたかもしれない。
(PRコメントで触れた)関数型スタイルで記述する恩恵はあまりなく、
~~ここは普通に for 文の方が解りやすいので変更します。~~

訂正: ここを変更するとコードに書いたコメントが全部ずれてしまったので差し戻します。
無駄に読みにくくなってしまっただけで、現状のままでも動作に支障ありません。

kagekiyo7 replied:
`.fromEntries`がSafari 12.1未満非対応なので、使用するのならindex.htmlに記載されている注意事項更新したほうがいいかもです。
(SafariはデフォルトアプリなのでiOSを更新する以外に更新する手段がない、恐らくSafariのバージョン=iOSのバージョン)
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries

>  使用しているライブラリが古いOSに対応していない為と思われますので、お手数ですがiOS10へのアップデートをお願い致します。

とりあえず僕のiPhone(iOS11)では非対応でホワイトアウトしましたが、統計見ると大半の人がiOS12以上であるみたいです。
![image](https://user-images.githubusercontent.com/22173829/72667866-c638b580-3a63-11ea-9cb5-10b1e57a1c66.png)

https://developer.apple.com/support/app-store/

kei-gbf replied:
safari は対象外だと思ってたけど、iOSのバージョンについて言及があったのですね。> index.html

古いブラウザ向けに babel の polyfill 設定するという方法もありますが、
.fromEntries を使わないといけないという事もないので、 for文に変更した方が良さそうです。

ただし、先程試したところ、副作用で、ここのコードに書いたコメントが大幅にずれてしまった為、
このPR (43) ではこのまま残して、後で修正してもらっても良いでしょうか。

```js
getInitialSwitchState: function (_switchList=switchList) {
    const res = {};
    for (const {items} of _switchList) {
        for (const {name, selected} of items) {
            res[name] = selected || false;
        }
    }
    return res;
},
```

kagekiyo7 replied:
了解しました。

kei-gbf commented on src/result.js:
```diff
@@ -219,33 +238,29 @@ var ResultList = CreateClass({
-        res.previousCombinations = null;
-        res.ruleMaxSize = true;
-        res.filterOptionsChanged = false;
-        return res;
```
res.xxx ～に変更すると、処理が冗長になる他、
スペルが間違っていないかの確認作業が発生するので、
（大抵はコピペ編集なので大丈夫なはずですが、
　第三者からはどのように編集したかわからない。ミスは有りえる）
可能なら元のコードを残す方が良いです。

kei-gbf commented on src/result.js:
```diff
@@ -219,33 +238,29 @@ var ResultList = CreateClass({
-                    res[item] = 0;
-                }
-            }
-        }
```
上のgetInitialSwitchState関数に相当しますが、

- デフォルト値を設定する際に `if (item ===` を増やすことになる
  - コードを変更するので、バグの際に他の項目へ影響しやすい
  - またレビューで差分を見る際、行が長くなり変更箇所が解りにくい
    ※ 環境によっては1行80文字の表示制限もあるので (79文字+改行文字)
         長くなる時は改行を挟みます。今回は代案(getInitialSwitchState関数)を提示。
- 殆どの項目は false ですが、item === と x4回チェックされる
  ロジックとしては一律 0 で初期化してから、必要な項目のみを上書きするほうが良い。
  → switchList に初期値情報を追加。

kei-gbf commented on src/result.js:
```diff
@@ -695,37 +682,22 @@ var ResultList = CreateClass({
-
-                        <DropdownButton title={intl.translate("キャラ情報・スキル合計値", locale)} id="chara-and-skill-info">
-                            {generateMenuItem("キャラ情報・スキル合計値", true)}
-                        </DropdownButton>
```
カテゴリーを追加・変更する際、複数個所編集しなければならないので、
自動生成。第２引数のフラグは省略してます。もし必要なら、switchList 側に情報を持たせる。

kei-gbf commented on src/result.js:
```diff
@@ -695,37 +682,22 @@ var ResultList = CreateClass({
+                    onClick={this.handleEvent.bind(this, name)}
+                    active={this.state[name]}>
+                    {intl.translate(label, locale)}
+                </MenuItem>);
```
関数コンポーネント化することで記述を簡単に。

kei-gbf commented on src/result.js:
```diff
@@ -525,45 +540,18 @@ var ResultList = CreateClass({
-                    }
-                }
-
-                return res;
```
ここの課題は、レイアウトを弄りたい時にこのコードに手を入れないといけない点。

- ３列に分割する処理は chunk関数で。(レイアウトロジックを分離)

kei-gbf replied:
コメントの参照範囲の設定ミス。
count や temp.push を用いているロジック部分です。

kei-gbf commented on src/result.js:
```diff
@@ -287,13 +302,13 @@ var ResultList = CreateClass({
     handleEvent: function (key, e) {
         var newState = this.state;
-        newState[key] = (newState[key] == 0) ? 1 : 0;
+        newState[key] = !newState[key];
```
boolean (true|false) にすることで、各所で == 0  のようなチェックを省けます。
null 等も許容するようになりますが、JavaScript では暗黙の型変換があるのでエラーにはなりません。
!null => true (他の言語ではダメなこともあるので注意)

kei-gbf replied:
※ 補足、これはPRへ対してのレビューではなく、
元からあるコードに対しての変更です。
コードが少し簡潔に書けるようになるので変更しました。

kei-gbf commented:
コードにコメント付けたのですが、PRからはコードの参照範囲が解り難いので
こちらを参照してください。Files Changed
https://github.com/kagekiyo7/motocal/pull/43/files

kei-gbf committed:
Remove unnessesary argument

kagekiyo7 commented:
PRと指摘&解説ありがとうございます。
リファクタリングの手腕に感服するばかりです…！
この後も何回か読み直したいと思います。

>※この変更により disableAutoResultUpdate のテスト も必要になります（影響範囲）

不具合という程でもないですが、自動更新OFF→ON時に更新が行われない(masterは行われる)という相違点がありました。

----
こちらも不具合と言うほどでもないのですが(むしろ使いやすくなってる?)、リストの項目名をクリックしてもリストが閉じないという相違点がありました。(masterは一回一回閉じる)

kei-gbf commented:
参考になれば幸いです。余談ですが、興味があれば手法について…軽く紹介
chunk関数や for文 → Array.map 書き換え等、今回の変更での肝となってるのは
Functional-Programming『関数型(函数)プログラミング』という
プログラミングスタイルに基づくもので、複雑なロジックを簡潔に記述できたりと
コードが単に短くなる以上の恩恵があったりします。

----
- [x] 自動更新off が効いてない不具合修正
- [x] リストの項目の意図しない挙動変更

> 自動更新OFF→ON時に更新が行われない(masterは行われる)という相違点がありました。

```js
// 変更前
newState[key] == 0
```

元のコードでは状態が 0 の時に実行なので、ここは逆になってしまってました。
恐らく、自動更新OFF が効いてない状態になってます。（沢山の武器を入力する際に困る）

```diff
-        if (key == "disableAutoResultUpdate" && newState[key]) {
+        if (key == "disableAutoResultUpdate" && !newState[key]) {
```

----
もう一つの方は元の挙動から確認してみますが、
PC版での振る舞いのみですか？モバイル版は使ってなかったので元の挙動わからない。
変更した覚えはないので、意図しない挙動変更。
念の為、原因は調べてみます。（団イベ挟んでしまうと時間かかります）

意図した変更で改善になってるならPRに書いたほうが良いけど、
(仮定の話ですが) もし閉じない挙動が不採用になって元に戻そうとしても、
何処で挙動を抑制してるのか把握してないので、
このPR毎取りやめることになってしまう事がリスク。

次善の策としては、PRコメントに注記だけして、問題があってから別PRで対処する。

kei-gbf committed:
fix disableAutoResultUpdate

It was caused by previous commits in this PR,
when flag state was changed from number to boolean.

kei-gbf committed:
Cleanup code style

- add a space between arguments

kei-gbf commented on src/result.js:
```diff
@@ -695,37 +682,22 @@ var ResultList = CreateClass({
+                            <DropdownButton title={intl.translate(label, locale)} id={"dropdown-"+ind} key={ind}>
+                                {items.map(({name, label}) => <ResultMenuItem key={name} name={name} label={label} />)}
+                            </DropdownButton>
+                        )}
```
NOTE: DropdownButton の挙動 (選択時に毎回閉じない)

挙動変更との関連は解りませんが、出力されるコードの相違点(メモ)
- id

kagekiyo7 commented:
PC版ですね。挙動比較のgifを貼っておきます(master→このPR)
![fdsadfsa2](https://user-images.githubusercontent.com/22173829/72666922-dcda0f00-3a59-11ea-9cb2-6676f9d89952.gif)

kei-gbf committed:
Change overused map to for-loop

the overused functional programming style had made tricky code,
no much other merits on this case now,
simple for-loop is easy to read.

kei-gbf commented:
すいません。
PRに書いたコードコメントが大幅にずれてしまったので、f48f232 取り消します。

kei-gbf committed:
Revert "Change overused map to for-loop"

This reverts commit f48f23260ce14ae7697f2a78376725db5f87983d.

kei-gbf commented:
比較ありがとうございます。確かに手間が減って使いやすそうになってる
他に弊害が出て無ければいいのですが。Dropdown 部分だけ取り出して試してみます。

- https://react-bootstrap-v3.netlify.com/components/dropdowns/
  ライブラリ側では multi-select には対応しておらず、motocal側で振る舞いを制御

kei-gbf commented:
リストが閉じない問題

逆にデメリットは、単項目のみ選択するケースが多い場合では
リストが勝手に閉じないので一手間増える。

簡単なデモを作り試してみましたが、ライブラリ側の振る舞いとしては
~~クリック時に勝手に閉じない様です。~~
https://codepen.io/kei-gbf/pen/povxYrV?editors=0111

kei-gbf commented:
変更前のコード generateMenuItem に、console.log を挿入して
どのタイミングで再描画されているのかを確かめてみましたが

以前の振る舞いは、項目が選択される度にリスト全体が再生成されていたようです。
本来の振る舞いであれば、項目選択時は選択された項目のみを再描画すればいいので、
codepen のデモの方では、選択時に選択した項目のみしかログを出力しません。

意図しない変更でしたが、他に差し支えなければ
ここは不具合修正扱いでも良いかもしれません。

もし選択時に自動的に閉じる振る舞いが必要でも、
~~親コンポーネントの再描画(再生成)の副作用として結果的に閉じているよりは、~~
再描画を行わず明示的に閉じた方が良い。

kagekiyo7 commented:
なるほど、元々はこういう挙動なのですね。了解しました。
ChromeのDevice ModeでのチェックですがスマホUIも問題なさそうです。

kei-gbf commented:
> 元々はこういう挙動なのですね

すいません訂正
- ~~ライブラリ側の振る舞いとしてはクリック時に勝手に閉じない様です。~~
- ~~親コンポーネントの再描画(再生成)の副作用として結果的に閉じている~~
  ここは推測でした。再描画に関しては先述の通り
  閉じない方では選択項目のみの再描画でしたが
  全て再描画しても閉じないケースを再現出来たので、閉じる閉じないとは直接関連しない

公式のデモでは選択時に閉じられました。
https://react-bootstrap-v3.netlify.com/components/dropdowns/

ライブラリが想定する用途としては、アプリのメニューなので
閉じるのが正しい挙動のようです。

ということで、自分で書いたデモでの使い方が間違っていた
本来閉じるはずが、イベントが正しく伝搬されてなかった可能性もある。

----
閉じる方法も再現出来ましたが、カスタマイズするとなると少し面倒そう。

URL内下の方の
Custom Dropdown Components に toggle 時のみ
開いたままにカスタマイズできるようですが、
他のライブラリを要求する等、手間のかかりそうな方法なので見送ります。

kei-gbf commented:
> イベントが正しく伝搬されてなかった可能性もある。

デバッガで追ってませんが、原因としてはこれが濃厚みたいです。

本来の挙動
- DropdownButton
  - MenuItem ↑クリック時 親コンポーネントを閉じる

試したデモ
- DropdownButton
  - MyMenuItem 直近の親が DropdownButton ではない為挙動が変わる
    - MenuItem  ↑クリック時 親コンポーネントを閉じる

コンポーネント化の副作用ですね。
該当箇所は、const ResultMenuItem 関数コンポーネントを利用している箇所。
```jsx
<ResultMenuItem key={name} name={name} label={label} />
```
ここのコンポーネント化を辞めることで従来の挙動になりました。
